internal/chain/reward.go

```diff
@@ -1,157 +1,312 @@
+		}
 	}
+	return payable, missing, nil
+}
+
+func ballotResult(
+	bc engine.ChainReader, header *block.Header, shardID uint32,
+) (shard.SlotList, shard.SlotList, shard.SlotList, error) {
 	// TODO ek – retrieving by parent number (blockNum - 1) doesn't work,
 	//  while it is okay with hash.  Sounds like DB inconsistency.
 	//  Figure out why.
 	parentHeader := bc.GetHeaderByHash(header.ParentHash())
 	if parentHeader == nil {
-		return ctxerror.New("cannot find parent block header in DB",
-			"parentHash", header.ParentHash())
+		return nil, nil, nil, ctxerror.New(
+			"cannot find parent block header in DB",
+			"parentHash", header.ParentHash(),
+		)
 	}
 	if parentHeader.Number().Cmp(common.Big0) == 0 {
 		// Parent is an epoch block,
 		// which is not signed in the usual manner therefore rewards nothing.
-		return nil
+		return shard.SlotList{}, shard.SlotList{}, shard.SlotList{}, nil
 	}
 	parentShardState, err := bc.ReadShardState(parentHeader.Epoch())
 	if err != nil {
-		return ctxerror.New("cannot read shard state",
-			"epoch", parentHeader.Epoch(),
+		return nil, nil, nil, ctxerror.New(
+			"cannot read shard state", "epoch", parentHeader.Epoch(),
 		).WithCause(err)
 	}
-	parentCommittee := parentShardState.FindCommitteeByID(parentHeader.ShardID())
+	parentCommittee := parentShardState.FindCommitteeByID(shardID)
 	if parentCommittee == nil {
-		return ctxerror.New("cannot find shard in the shard state",
+		return nil, nil, nil, ctxerror.New(
+			"cannot find shard in the shard state",
 			"parentBlockNumber", parentHeader.Number(),
 			"shardID", parentHeader.ShardID(),
 		)
 	}
-	var committerKeys []*bls.PublicKey
-	for _, member := range parentCommittee.Slots {
-		committerKey := new(bls.PublicKey)
-		err := member.BlsPublicKey.ToLibBLSPublicKey(committerKey)
-		if err != nil {
-			return ctxerror.New("cannot convert BLS public key",
-				"blsPublicKey", member.BlsPublicKey).WithCause(err)
-		}
-		committerKeys = append(committerKeys, committerKey)
-	}
-	mask, err := bls2.NewMask(committerKeys, nil)
-	if err != nil {
-		return ctxerror.New("cannot create group sig mask").WithCause(err)
+	payable, missing, err := blockSigners(parentHeader, parentCommittee)
+	return parentCommittee.Slots, payable, missing, err
+}
+
+func ballotResultBeaconchain(
+	bc engine.ChainReader, header *block.Header,
+) (shard.SlotList, shard.SlotList, shard.SlotList, error) {
+	return ballotResult(bc, header, shard.BeaconChainShardID)
+}
+
+// AccumulateRewards credits the coinbase of the given block with the mining
+// reward. The total reward consists of the static block reward and rewards for
+// included uncles. The coinbase of each uncle block is also rewarded.
+func AccumulateRewards(
+	bc engine.ChainReader, state *state.DB, header *block.Header,
+	rewarder reward.Distributor, slasher slash.Slasher,
+	beaconChain engine.ChainReader,
+) error {
+
+	blockNum := header.Number().Uint64()
+	if blockNum == 0 {
+		// Epoch block has no parent to reward.
```

Comment:
you mean genesis block?

Reply:
this is original comment from code, will update to `genesis block `